cpp/ql/test/query-tests/Critical/UnsafeUseOfThis/test.cpp

```diff
@@ -0,0 +1,50 @@
+struct A { virtual void f() = 0; };
+
+struct B;
+void call_f(B*);
+
+struct B : public A {
+	B() {
+		call_f(this);
+	}
+
+    B(B& b) {
+        b.f(); // BAD: undefined behavior
+    }
+
+	~B() {
+		f(); // BAD: undefined behavior
+	}
+};
+
+struct C : public B {
+	C(bool b) {
+		call_f(this);
+
+		if(b) {
+			this->f(); // GOOD: Not a 'must' flow
+		}
+	}
+};
+
+struct D : public B {
+    D() : B(*this) {}
+};
+
+void call_f(B* x) {
+	x->f(); // 2 x BAD: Undefined behavior
+}
+
+struct E : public A {
+	E() {
+		f(); // GOOD: Will call `E::f`
+	}
+
+	void f() override {}
+};
+
+struct F : public E {
+	F() {
+		((A*)this)->f(); // BAD: undefined behavior
```

Comment:
I don't understand why this is `BAD`. Won't the program call `E::f` here? How is it undefined behaviour?

Reply:
On Clang, GCC, and MSVC it calls `E::f`, yes. It's undefined behavior since it's relying on the virtual call mechanism in a constructor (Both Clang and GCC report this case as undefined when compiling it).
If we want to avoid flagging such cases we could require that there is no override of `f` in any of `F`'s base classes.

Reply:
The relevant standard text is (from `[class.cdtor]`):
> For an object with a non-trivial constructor, referring to any non-static member or base class of the object
before the constructor begins execution results in undefined behavior. For an object with a non-trivial
destructor, referring to any non-static member or base class of the object after the destructor finishes
execution results in undefined behavior

Reply:
> referring to any non-static member or base class of the object before the constructor begins execution results in undefined behavior

We're referring to `f` _after_ the constructor begins execution, not _before_.

Reply:
Oh. I was reading the wrong section >.< The relevant quote paragraph 4 in `[class.cdtor]`:

> Member functions, including virtual functions (10.3), can be called during construction or destruction (12.6.2).
<b>When a virtual function is called directly or indirectly from a constructor</b> or from a destructor, including
during the construction or destruction of the class’s non-static data members, and the object to which the
call applies is the object (call it x) under construction or destruction, <b>the function called is the final overrider
in the constructor’s or destructor’s class and not one overriding it in a more-derived class.</b> If the virtual
function call uses an explicit class member access (5.2.5) and the object expression refers to the complete
object of x or one of that object’s base class subobjects but not x or one of its base class subobjects, the
behavior is undefined.

That makes it sound like Clang and GCC are wrong, and that the call is actually well-defined.

Reply:
Yes, it sounds like they're compiling the code correctly but giving a false warning.

Reply:
I've fixed the annotation in 0a6a22562b5166cc53398625a271a5084e41214a.